### PR TITLE
adding test case to token update precompile

### PR DESCRIPTION
**Pull request by tdermendzhievv**
**Description**:
This PR modifies token update precompile suite. It adds scenario where we are trying to update token treasury without treasury signature.
**Related issue(s)**:

Fixes #

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/codec/KeyValueWrapper.java [hunk 1]
```diff
@@ -163,7 +163,8 @@ public Key asGrpc() {
             case DELEGATABLE_CONTRACT_ID -> Key.newBuilder()
                     .setDelegatableContractId(delegatableContractID)
                     .build();
-            default -> throw new InvalidTransactionException(ResponseCodeEnum.FAIL_INVALID);
+            default -> throw new InvalidTransactionException(
+                    "INVALID_KEY", ResponseCodeEnum.FAIL_INVALID);
```
Comment (stoqnkpL) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/codec/KeyValueWrapper.java:167]: I think adding the detailed message "INVALID_KEY" is an improvement. Should we also change the response code to not be `FAIL_INVALID`? `FAIL_INVALID` is reserved for internal failures we're unable to recover from that may bring the whole node down. I think we can use something like `INVALID_TRANSACTION_BODY` here like we do in [verifySolidityInput()](https://github.com/hashgraph/hedera-services/blob/c107510646461121799ccefd036b465f0a809aee/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/impl/TokenCreatePrecompile.java#L472).

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/impl/TokenCreatePrecompile.java [hunk 2]
```diff
@@ -539,6 +525,24 @@ private void verifySolidityInput() {
...
+                final var tokenKeyBitField = tokenKey.keyType();
+                validateTrue(
+                        tokenKeyBitField != 0 && tokenKeyBitField < 128, INVALID_TRANSACTION_BODY);
+                for (int j = i + 1; j < tokenKeysSize; j++) {
+                    validateTrue(
+                            (tokenKeyBitField & tokenKeys.get(j).keyType()) == 0,
+                            INVALID_TRANSACTION_BODY);
+                }
+            }
+        }
```
Comment (stoqnkpL) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/impl/TokenCreatePrecompile.java:543]:
```suggestion
        if (tokenKeys.isEmpty()) {
          return;
        }
        for (int i = 0, tokenKeysSize = tokenKeys.size(); i < tokenKeysSize; i++) {
            final var tokenKey = tokenKeys.get(i);
            validateTrue(
                    tokenKey.key().getKeyValueType() != INVALID_KEY, INVALID_TRANSACTION_BODY);
            final var tokenKeyBitField = tokenKey.keyType();
            validateTrue(
                    tokenKeyBitField != 0 && tokenKeyBitField < 128, INVALID_TRANSACTION_BODY);
            for (int j = i + 1; j < tokenKeysSize; j++) {
                validateTrue(
                        (tokenKeyBitField & tokenKeys.get(j).keyType()) == 0,
                        INVALID_TRANSACTION_BODY);
            }
        }
```

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/impl/TokenUpdatePrecompile.java [hunk 3]
```diff
@@ -113,9 +125,65 @@ public void run(MessageFrame frame) {
...
+                            frame,
+                            treasuryId.asEvmAddress(),
+                            sigsVerifier::hasActiveKey,
+                            ledgers,
+                            aliases);
+
+            validateTrue(
+                    treasuryHasSigned,
+                    INVALID_FULL_PREFIX_SIGNATURE_FOR_PRECOMPILE,
+                    "Token update");
```
Comment (stoqnkpL) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/impl/TokenUpdatePrecompile.java:142]: Should we add a more meaningful message? Like "Treasury account signature missing in token update?"?

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/impl/TokenUpdatePrecompile.java [hunk 4]
```diff
@@ -113,9 +125,65 @@ public void run(MessageFrame frame) {
...
+
+        updateOp.tokenKeys().stream()
+                .filter(TokenKeyWrapper::isUsedForAdminKey)
+                .findFirst()
+                .ifPresent(
+                        key ->
+                                validateTrue(
+                                        validateAdminKey(frame, key),
+                                        INVALID_FULL_PREFIX_SIGNATURE_FOR_PRECOMPILE,
+                                        "Token update"));
```
Comment (stoqnkpL) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/impl/TokenUpdatePrecompile.java:153]: same as above, we can add a more meaningful message here

**Review by stoqnkpL — COMMENTED** [review]
Nice findings! I like these improvements!
Let us think about the handling of exceptions during precompile calls input decoding. 
Currently, invalid solidity inputs cause exceptions to be thrown in body() which are then caught in [decodeInput](https://github.com/hashgraph/hedera-services/blob/2b87d46a322c0cc1eda8233d818acf29573b727c/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/HTSPrecompiledContract.java#L796) and shadowed. There are a couple of issues here:
1. The end users never receive information about how their calls to the system contract failed.
2. We're logging a warning message with the exception in services, which is not the desired behavior (https://github.com/hashgraph/hedera-services/issues/4416)

I suggest removing the logging of the exception and looking into ways to propagate the detailed information about how the call failed to the caller. Maybe we can add the information to [the exceptional halt reason we're adding to the message frame in case the `transactionBody` is null](https://github.com/hashgraph/hedera-services/blob/2b87d46a322c0cc1eda8233d818acf29573b727c/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/HTSPrecompiledContract.java#L229)?

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/impl/TokenUpdatePrecompile.java [hunk 5]
```diff
@@ -113,6 +125,39 @@ public void run(MessageFrame frame) {
...
+                            frame,
+                            treasuryId.asEvmAddress(),
+                            sigsVerifier::hasActiveKey,
+                            ledgers,
+                            aliases);
+
+            validateTrue(
+                    treasuryHasSigned,
+                    INVALID_FULL_PREFIX_SIGNATURE_FOR_PRECOMPILE,
+                    "Treasury account signature missing in token update!");
```
Comment (IvanKavaldzhiev) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/impl/TokenUpdatePrecompile.java:142]: We can extract a constant with the custom error message.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/impl/TokenUpdatePrecompile.java [hunk 6]
```diff
@@ -113,6 +125,39 @@ public void run(MessageFrame frame) {
...
+                                validateTrue(
+                                        validateAdminKey(
+                                                frame,
+                                                key,
+                                                senderAddress,
+                                                sigsVerifier,
+                                                ledgers,
+                                                aliases),
+                                        INVALID_FULL_PREFIX_SIGNATURE_FOR_PRECOMPILE,
+                                        "New admin account signature missing in token update!"));
```
Comment (IvanKavaldzhiev) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/impl/TokenUpdatePrecompile.java:159]: We can extract a constant with the custom error message.

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/TokenUpdatePrecompileSuite.java [hunk 7]
```diff
@@ -108,6 +109,8 @@ public class TokenUpdatePrecompileSuite extends HapiSuite {
     private static final String DELEGATE_KEY = "tokenUpdateAsKeyDelegate";
     private static final String ACCOUNT_TO_ASSOCIATE = "account3";
     private static final String ACCOUNT_TO_ASSOCIATE_KEY = "associateKey";
+    public static final String UPDATE_TOKEN_WITH_ALL_FIELDS = "updateTokenWithAllFields";
+    public static final String UPDATE_TOKEN_TREASURY = "updateTokenTreasury";
     private static final String CUSTOM_NAME = "customName";
```
Comment (IvanKavaldzhiev) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/TokenUpdatePrecompileSuite.java:114]: Probably these fields should be `private`?

**Pull request closed without merging**